crates/shared/src/price_estimation/native/oneinch.rs

```diff
@@ -0,0 +1,237 @@
+use {
+    super::{NativePriceEstimateResult, NativePriceEstimating},
+    crate::price_estimation::{
+        Estimate,
+        PriceEstimateResult,
+        PriceEstimating,
+        PriceEstimationError,
+        Query,
+    },
+    anyhow::{anyhow, Result},
+    futures::{future::BoxFuture, FutureExt},
+    model::order::OrderKind,
+    number::nonzero::U256 as NonZeroU256,
+    primitive_types::{H160, U256},
+    reqwest::Client,
+    std::{
+        collections::HashMap,
+        sync::{Arc, Mutex},
+        time::Duration,
+    },
+};
+
+const BASE_URL: &str = "https://api.1inch.dev/";
+const REFRESH_INTERVAL: Duration = Duration::from_secs(12);
+
+struct OneInch {
+    client: Client,
+    chain_id: u64,
+    // Denominated in wei
+    prices: Arc<Mutex<HashMap<H160, U256>>>,
+    native_price_estimation_amount: NonZeroU256,
+    native_token: H160,
+}
+
+impl OneInch {
+    #[allow(dead_code)]
+    pub fn new(
+        client: Client,
+        chain_id: u64,
+        native_price_estimation_amount: NonZeroU256,
+        native_token: H160,
+    ) -> Self {
+        let instance = Self {
+            client,
+            chain_id,
+            prices: Arc::new(Mutex::new(HashMap::new())),
+            native_price_estimation_amount,
+            native_token,
+        };
+        instance.update_prices_in_background();
+        instance
+    }
+
+    fn update_prices_in_background(&self) {
+        let client = self.client.clone();
+        let chain = self.chain_id;
+        let prices = self.prices.clone();
+        tokio::task::spawn(async move {
+            loop {
+                match update_prices(&client, chain).await {
+                    Ok(new_prices) => {
+                        *prices.lock().unwrap() = new_prices;
+                    }
+                    Err(err) => {
+                        tracing::warn!(?err);
+                    }
+                }
+                tokio::time::sleep(REFRESH_INTERVAL).await;
+            }
+        });
+    }
+}
+
+impl NativePriceEstimating for OneInch {
+    fn estimate_native_price<'a>(
+        &'a self,
+        token: &'a H160,
```

Comment:
I overlooked this when I simplified the price APIs but it makes more sense for `token` to be passed by value because it's already a `Copy` type. 🤦‍♂️

Reply:
Will refactor separately